dna/modify.go

```diff
@@ -133,6 +150,22 @@ func Insert(seq []Base, insPos int64, insSeq []Base) []Base {
 	return append(seq[:insPos], append(insSeq, seq[insPos:]...)...)
 }
 
+// InsertStable is identical to Insert, but conserves the
+// underlying memory of the input destSeq slice if possible
+func InsertStable(destSeq *[]Base, insPos int, insSeq []Base) error {
+	if insPos < 0 || insPos > len(*destSeq) {
+		return ErrInvalidPosition
+	}
+	if len(*destSeq)+len(insSeq) > cap(*destSeq) {
+		*destSeq = Insert(*destSeq, int64(insPos), insSeq)
+		return ErrExceedCapNewSliceCreated
+	}
+	*destSeq = (*destSeq)[:len(*destSeq)+len(insSeq)]                                    // extend the LEN of the slice to include the inserted bases
+	copy((*destSeq)[insPos+len(insSeq):cap(*destSeq)], (*destSeq)[insPos:cap(*destSeq)]) // shift bases in slice by len(insSeq) at the position of the insertion
```

Comment:
I think this is copying more than you need to.  You only need to copy the bases from the original seq that come after the insertion:
copy((*destSeq)[insPos+len(insSeq):len(*destSeq)], (*destSeq)[insPos:len(*destSeq)-len(insSeq)])

Reply:
Craig, when I follow that code it looks like the number of slices is staying the same. Wouldnt subtracting the len(insSeq) defeat the purpose of increase by the length of the insertion in the destination argument?
assuming: copy (dst, src)
I made an example to help me think through these with destSeq [1:100] insPos = 50 and len(insSeq) = 10
With Dan's code I get: 
copy (60:100, 50:100) 
but with your code I'm getting:
copy(60:100, 50:90). 
Maybe I'm thinking about this all wrong as we can chat during code review?

Reply:
I think the trick is that Dan is re-sizing the slice on the line above, so when the copy is executed, the length is no longer 100, but 110:
destSeq is originally a slice of length 100 (index 0 to index 99)
We are going to insert a 10bp, so now we change it to a slice of length 110, where the first 100 bases are the starting seq and the last 10 are random values
Then the copy command says take the values in destSeq[50 to (110-10)] aka destSeq[50 to 100] and put them in destSeq[50+10 to 110] aka destSeq[60 to 110]. This serves to "shift" all the bases in the original sequence that come after the insertion, to make room for the insertion.